osu.Game.Rulesets.Mania/Mods/ManiaModConstant.cs

```diff
@@ -20,30 +21,38 @@ public class ManiaModConstant : Mod, IApplicableToBeatmap
 
         public void ApplyToBeatmap(IBeatmap beatmap)
         {
-            // Get all difficulty points
-            var difficultyPoints = beatmap.ControlPointInfo.DifficultyPoints;
-            // Get all timing points
-            var timingPoints = beatmap.ControlPointInfo.TimingPoints;
             // Get the initial BPM
-            var initialBPM = timingPoints[0].BPM;
+            var initialBPM = beatmap.ControlPointInfo.TimingPoints[0].BPM;
 
-            foreach (var difficultyPoint in difficultyPoints)
+            foreach (var group in beatmap.ControlPointInfo.Groups)
             {
-                // Set this difficulty point's speed multiplier to 1
-                difficultyPoint.SpeedMultiplier = 1;
-            }
+                // Get the timing control point of this group, if any
+                var timingControlPoints = group.ControlPoints.OfType<TimingControlPoint>().ToArray();
 
-            // Correct changes in speed caused by timing points by adding an associated difficulty point
-            foreach (var timingPoint in timingPoints)
-            {
-                // Create a new difficulty point to counteract timing point
-                DifficultyControlPoint diffControlPoint = new DifficultyControlPoint
+                // Get the difficulty control point of this group, if any
+                var difficultyControlPoints = group.ControlPoints.OfType<DifficultyControlPoint>().ToArray();
+
+                // If there is a timing control point
+                if (timingControlPoints.Any())
+                {
+                    var timingControlPoint = timingControlPoints[0];
+
+                    DifficultyControlPoint diffControlPoint = new DifficultyControlPoint();
+
+                    diffControlPoint.SpeedMultiplierBindable.MinValue = double.MinValue; // Uncapped minimum value
+                    diffControlPoint.SpeedMultiplierBindable.MaxValue = double.MaxValue; // Uncapped maximum value
+                    diffControlPoint.SpeedMultiplierBindable.Precision = 0.000000001d; // Arbitrary value, may need changing
+                    diffControlPoint.SpeedMultiplier = initialBPM / timingControlPoint.BPM; // Counteract BPM velocity
+
+                    group.Add(diffControlPoint);
+                }
+                // Else, there is no timing control point, but a difficulty control point
+                else if (difficultyControlPoints.Any())
```

Comment:
Not sure this is correct - I don't think there's anything stopping a group from having a timing point _and_ a difficulty control point? Should just probably be

```suggestion
                if (difficultyControlPoints.Any())
```

Reply:
If it were to check for a difficulty point after it checks for a timing point, there's a possibility that it will remove the difficulty point, which will undo the timing point's counteraction. From my understanding, at least. That's why it's an `else if`. If the group has both a timing and difficulty point in the first place, the difficulty point will be overridden anyway.

Reply:
Oh, because there's a difficulty point potentially added before... Yeah I think that's a bit backwards. Should clean up the group from any difficulty points first before proceeding to add own overrides here I reckon.